### PR TITLE
Fixed bugs in the async journal reader

### LINKED ISSUE #28: async add_filter coroutine never awaited
This code make this issue :
```python3
            reader = AsyncJournalReader()
            await reader.open(JournalOpenMode.SYSTEM)
            await reader.seek_tail()
            await reader.add_filter((
                Rule("_SYSTEMD_UNIT", "some-service.service")
            ))

RuntimeWarning: coroutine 'Base._exec' was never awaited
  Rule("_SYSTEMD_UNIT", "some-service.service")
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

```

Solution:

an `await` has to be added to the `add_filter` method from `AsynJournalReader` class.

from:
```python3
    async def add_filter(self, rule):
        return self._exec(self.__reader.add_filter, rule)
```

to:
```python3
    async def add_filter(self, rule):
        return await self._exec(self.__reader.add_filter, rule)
```

### PR DESCRIPTION
**Pull request by agronholm**
* Added missing "await" statements to several methods
* Implemented the next() method
* Replaced async iterator implementation (the old one got stuck waiting on an event after the synchronous iterator had already exited)

Fixes #28.

### CODE REVIEW

**Comment by mosquito** [comment]
Thank you for your contribution, unfortunately I worked on similar fixes last week, after my tests I released `cysystemd==1.4.0` and it's conflicts with this changeset.

**Comment by agronholm** [comment]
No problem, so long as the problems get fixed. I should take a look at your new code and see if it checks out.

**Comment by mosquito** [comment]
@agronholm actually you third case not fixed yet. I'm trying to fix it without removing `AsyncReaderIterator` right now.

**Comment by agronholm** [comment]
You can use the same approach there. I can make a PR if you like.

**Comment by mosquito** [comment]
@agronholm I fixed it [here](https://github.com/mosquito/cysystemd/pull/30/files) I will happy if you could test it on your cases

**Comment by agronholm** [comment]
Why did you add a `.pyi` file? Why not just add the type annotations to the code itself? It's not like Python < 3.5 can parse the code anyway with all those `await` statements there.

**Comment by agronholm** [comment]
Tell me what you think of my alternative fix. I'm going to bed now.

File: cysystemd/async_reader.py [hunk 1]
```diff
@@ -133,30 +126,53 @@ def skip_next(self, skip):
...
+        self._loop.create_task(self._exec(self._reader))
+        self.closed = False
 
-    async def __aiter__(self):
-        def read_entries():
-            for item in self.__reader:
-                asyncio.run_coroutine_threadsafe(queue.put(item), self._loop)
+    def _reader(self):
+        for item in self.reader:
+            asyncio.run_coroutine_threadsafe(self.queue.put(item), self._loop)
```
Comment (mosquito) [inline cysystemd/async_reader.py:165]: The future will be created and will be never awaited. That's means queue not been overflow, but event loop will await created tasks, as result memory will be allocated and journal will be read but queue will newer no consumed.

File: cysystemd/async_reader.py [hunk 2]
```diff
@@ -133,30 +126,53 @@ def skip_next(self, skip):
...
 
-            asyncio.run_coroutine_threadsafe(queue.put(None), self._loop)
+        self.closed = True
+        asyncio.run_coroutine_threadsafe(self.queue.put(None), self._loop)
```
Comment (mosquito) [inline cysystemd/async_reader.py:168]: The "tail" of created and not awaited coroutines in the loop will schedule not sequentially and this "flag" might be send not last.

Comment (agronholm) [reply cysystemd/async_reader.py:168]: You bring up a good point – I am ashamed I didn't think of this!

Comment (agronholm) [reply cysystemd/async_reader.py]: Ok, I've fixed that problem now.

**Comment by agronholm** [comment]
The simplest way to iterate it would be this:
```python3
    # ... in the AsyncJournalReader class
    def __aiter__(self):
        return self

    async def __anext__(self):
        try:
            return await self._loop.run_in_executor(self._executor, self.__reader.next)
        except StopIteration:
            raise StopAsyncIteration
```
Do you really intend to do something about multiple concurrent iterators? Or raise an error when that is being tried?

**Comment by agronholm** [comment]
Waiting on a synchronous lock is admittedly not very good, but I also think that there is no need for a separate iterator instance.

**Comment by agronholm** [comment]
Hmm – am I understanding this correctly? If I do `for entry in reader:` it will read entries in a non-blocking fashion until there are no entries anymore, and then we need to wait on the reader to get more entries?

**Comment by mosquito** [comment]
Oh... sorry I merged the wrong PR. Master has been force updated, if you still want to fix it please create a new one against master.

**Pull request merged**